[@types/node] Allow to extends Http2ServerResponse and MyHttp2ServerRequest

chyzwar opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

This makes constructor public for Http2ServerRequest and Http2ServerResponse. According to http2 docs, it is possible to extend these classes. 

@see https://nodejs.org/api/http2.html#http2_http2_createsecureserver_options_onrequesthandler

There will be a need for more changes in http2 to fully support custom req, res classes.

Flarna commented on types/node/http2.d.ts:
```diff
@@ -583,7 +583,7 @@ declare module "http2" {
 
     export class Http2ServerRequest extends stream.Readable {
-        private constructor();
+        constructor();
```
The arguments are missing, see https://github.com/nodejs/node/blob/0481a7f3b883862cadf963314d23490adc6656a4/lib/internal/http2/compat.js#L274
As long as they were private it was not needed to specify the arguments as it was anyway not possible to call it but if you extend it you have to pass the correct args here.

chyzwar replied:
This is now done.

Flarna commented on types/node/http2.d.ts:
```diff
@@ -651,7 +651,7 @@ declare module "http2" {
 
     export class Http2ServerResponse extends stream.Stream {
-        private constructor();
+        constructor();
```
The arguments are missing, see https://github.com/nodejs/node/blob/0481a7f3b883862cadf963314d23490adc6656a4/lib/internal/http2/compat.js#L428

chyzwar replied:
I added type for the first argument but second "options" I am not sure what should be used. It seems options are passed to superclass Stream. Stream class do not define specific constructor options ?

chyzwar replied:
it is bug in node? this constructor should be empty ?

Flarna replied:
why empty? see https://github.com/nodejs/node/blob/0481a7f3b883862cadf963314d23490adc6656a4/lib/internal/http2/compat.js#L743 the first argument is a `ServerHttp2Stream`. I think the second arg can be omitted.

chyzwar replied:
It is already like that. Thanks for confirming the second arg.

chyzwar commented:
I am will look into tests but I am not sure if these would complete without making Http2Server generic.

Flarna commented on types/node/http2.d.ts:
```diff
@@ -583,7 +583,7 @@ declare module "http2" {
 
     export class Http2ServerRequest extends stream.Readable {
-        private constructor();
+        constructor(stream: ServerHttp2Stream, headers: string[], options: stream.ReadableOptions, rawHeaders: string);
```
I think `headers` should be of type `IncomingHttpHeaders` and `rawHeaders` should be `string[]`. According to source `options` may be also `undefined`.

chyzwar replied:
thanks, fixed.

Flarna commented:
> I am will look into tests but I am not sure if these would complete without making Http2Server generic.

You could use the test in http as sample.
I agree that a full solution (correct typed events,...) would require a generic `Http2Server` but as far as I remember this is not really possible with typescript 2.1. At least I failed to do this for http; I think it was because of missing default types for generics (and maybe something else).

chyzwar commented:
Thanks for review. Added tests for custom req, res classes. I added them as a separate section. I will try to make generic Http2Server next PR and If successful I will add tests for requestHandler.

chyzwar commented:
Thank you everyone.
